models/ellipsis_finance/bnb/ellipsis_finance_bnb_pool_tokens.sql

```diff
@@ -0,0 +1,380 @@
+            ('0x19ec9e3f7b21dd27598e7ad5aae7dc0db00a806d', '1', '0xA6fDEa1655910C504E974f7F1B520B74be21857B'), 
+            ('0x19ec9e3f7b21dd27598e7ad5aae7dc0db00a806d', '2', '0x5f7f6cB266737B89f7aF86b30F03Ae94334b83e9'), 
+        -- valbtc_renbtc 
+            ('0xfa715e7c8fa704cf425dd7769f4a77b81420fbf2', '0', '0x204992f7fCBC4c0455d7Fec5f712BeDd98E7d6d6'), 
+            ('0xfa715e7c8fa704cf425dd7769f4a77b81420fbf2', '1', '0xfCe146bF3146100cfe5dB4129cf6C82b0eF4Ad8c'),
+        -- valtusd_val3ps
+            ('0xAB499095961516f058245C1395f9c0410764b6Cd', '0', '0xBB5DDE96BAD874e4FFe000B41Fa5E98F0665a4BC'), 
+            ('0xAB499095961516f058245C1395f9c0410764b6Cd', '1', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5'), -- duplicate for both pools because there's token id 2 & 3 in exchange underlying 
+            ('0xAB499095961516f058245C1395f9c0410764b6Cd', '2', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5'),
+            ('0xAB499095961516f058245C1395f9c0410764b6Cd', '3', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5'), 
+        -- valdai_val3ps
+            ('0x245e8bb5427822FB8fd6cE062d8dd853FbcfABF5', '0', '0x2c85EBAE81b7078Cd656b2C6e2d58411cB41D91A'),
+            ('0x245e8bb5427822FB8fd6cE062d8dd853FbcfABF5', '1', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5'),
+            ('0x245e8bb5427822FB8fd6cE062d8dd853FbcfABF5', '2', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5'),
+            ('0x245e8bb5427822FB8fd6cE062d8dd853FbcfABF5', '3', '0x5b5bD8913D766D005859CE002533D4838B0Ebbb5')
+        ) as temp_table (pool, token_id, token_address)
+), 
+
+base_pools_underlying_tokens_sold as ( -- the underlying tokens sold works different for base pools so we're using duplicate the pool tokens here
+        SELECT 
+            pool, 
+            token_id, 
+            token_address, 
+            'underlying_token_sold' as token_type, 
+            pool_type
+        FROM 
+        base_pools_pool_tokens
+), 
+
+plain_pools as ( -- getting plain pools data 
+        SELECT 
+            pool, 
+            CAST(token_id AS STRING) as token_id,
+            CAST(token_address AS STRING) as token_address,
+            token_type
+        FROM 
+        (
+            SELECT
+                pool,
+                POSEXPLODE(coins) as (token_id, token_address),
+                'pool_token' as token_type
+            FROM
+            {{ source('ellipsis_finance_bnb', 'FactoryPool_evt_PlainPoolDeployed') }}
+            {% if is_incremental() %}
+            WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+            {% endif %}
+
+            UNION
+
+            SELECT
+                pool,
+                POSEXPLODE(coins) as (token_id, token_address),
+                'pool_token' as token_type
+            FROM
+            {{ source('ellipsis_finance_bnb', 'FactoryPool_v2_evt_PlainPoolDeployed') }}
+            {% if is_incremental() %}
+            WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
```

Comment:
Great work @henrystats, just one question, do you think any `UNION` in this .sql file can be replaced by `UNION ALL`? Like for `UNION` in those `x` tables.

Reply:
hmm sorry I thought I commented somewhere in the query but there were cases of pools (factory pools) being deployed multiple times so I'm assuming there was some redeployment somewhere, but since each pool would have it's own unique address then I think we can change to UNION ALL.

Just not for the factory pools because they were deployed multiple times.

Reply:
Yes you commented here.
https://github.com/duneanalytics/spellbook/pull/2260/files#diff-97198bcff111f20148dd2807d972e687826a4539f8d61e499aea49aaf3f2d2a5R274

`UNION` is a safer option, but if we having downstream CTE with `UNION`, we may have `UNION ALL` in upstream? As downstream `UNION` solves duplicates in one place?

Reply:
ah sorry I'm a bit confused now, what changes do I need to make?

Reply:
One `UNION` in the final stage may solve the repeated record problem, here is a demo:
- https://dune.com/queries/1836418
- https://dune.com/queries/1836445
- https://dune.com/queries/1836503
All three queries having the same number of row count(485 rows)